File: 3100-3199/3121. Count the Number of Special Characters II/Solution.java
//Approach-1 (Array Tracking)
//T.C : O(n)
//S.C : O(1)

class Solution {

  public int numberOfSpecialChars(String word) {

    int[] lowerLast = new int[26];

    int[] upperFirst = new int[26];

    Arrays.fill(lowerLast, -1);

    Arrays.fill(upperFirst, -1);

    for (int i = 0; i < word.length(); i++) {

      char ch = word.charAt(i);

      if (Character.isLowerCase(ch)) {

        lowerLast[ch - 'a'] = i;
      } else {

        if (upperFirst[ch - 'A'] == -1) {

          upperFirst[ch - 'A'] = i;
        }
      }
    }

    int count = 0;

    for (int i = 0; i < 26; i++) {

      if (lowerLast[i] != -1 &&
          upperFirst[i] != -1 &&
          lowerLast[i] < upperFirst[i]) {

        count++;
      }
    }

    return count;
  }
}

/*
 * //Approach-2 (HashMap Tracking)
 * //T.C : O(n)
 * //S.C : O(n)
 * 
 * class Solution {
 * 
 * public int numberOfSpecialChars(String word) {
 * 
 * HashMap<Character, Integer> firstUpper = new HashMap<>();
 * 
 * HashMap<Character, Integer> lastLower = new HashMap<>();
 * 
 * int c = 0;
 * 
 * for (int i = 0; i < word.length(); i++) {
 * 
 * char ch = word.charAt(i);
 * 
 * if (Character.isLowerCase(ch)) {
 * 
 * lastLower.put(ch, i);
 * }
 * else {
 * 
 * if (!firstUpper.containsKey(ch)) {
 * 
 * firstUpper.put(ch, i);
 * }
 * }
 * }
 * 
 * for (char ch = 'a'; ch <= 'z'; ch++) {
 * 
 * char upper = Character.toUpperCase(ch);
 * 
 * if (lastLower.containsKey(ch) &&
 * firstUpper.containsKey(upper)) {
 * 
 * if (lastLower.get(ch) < firstUpper.get(upper)) {
 * 
 * c++;
 * }
 * }
 * }
 * 
 * return c;
 * }
 * }
 */

File: 3100-3199/3121. Count the Number of Special Characters II/Solution.cpp
// Approach-1 (Array Tracking)
// T.C : O(n)
// S.C : O(1)

class Solution
{
public:
  int numberOfSpecialChars(string word)
  {

    vector<int> lowerLast(26, -1);

    vector<int> upperFirst(26, -1);

    for (int i = 0; i < word.length(); i++)
    {

      char ch = word[i];

      if (islower(ch))
      {

        lowerLast[ch - 'a'] = i;
      }
      else
      {

        if (upperFirst[ch - 'A'] == -1)
        {

          upperFirst[ch - 'A'] = i;
        }
      }
    }

    int count = 0;

    for (int i = 0; i < 26; i++)
    {

      if (lowerLast[i] != -1 &&
          upperFirst[i] != -1 &&
          lowerLast[i] < upperFirst[i])
      {

        count++;
      }
    }

    return count;
  }
};

/*
//Approach-2 (HashMap Tracking)
//T.C : O(n)
//S.C : O(n)

class Solution {
public:

    int numberOfSpecialChars(string word) {

        unordered_map<char, int> firstUpper;

        unordered_map<char, int> lastLower;

        int c = 0;

        for (int i = 0; i < word.length(); i++) {

            char ch = word[i];

            if (islower(ch)) {

                lastLower[ch] = i;
            }
            else {

                if (!firstUpper.count(ch)) {

                    firstUpper[ch] = i;
                }
            }
        }

        for (char ch = 'a'; ch <= 'z'; ch++) {

            char upper = toupper(ch);

            if (lastLower.count(ch) &&
                firstUpper.count(upper)) {

                if (lastLower[ch] < firstUpper[upper]) {

                    c++;
                }
            }
        }

        return c;
    }
};
*/